src/server/scene/session_locker.cpp

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "session_locker.h"
+#include "mir/frontend/surface_stack.h"
+#include "mir/console_services.h"
+#include "mir/fatal.h"
+
+#include <algorithm>
+
+namespace ms = mir::scene;
+
+ms::SessionLocker::SessionLocker(
+    std::shared_ptr<mf::SurfaceStack> const& surface_stack)
+    : surface_stack{surface_stack}
+{
+}
+
+void ms::SessionLocker::add_observer(std::shared_ptr<mf::SessionLockObserver> const& observer)
+{
+    observers.push_back(observer);
+}
+
+void ms::SessionLocker::remove_observer(std::weak_ptr<mf::SessionLockObserver> const& observer)
+{
+    observers.erase(std::remove(observers.begin(), observers.end(), observer.lock()));
+}
+
+void ms::SessionLocker::request_lock()
+{
+    if (!surface_stack->screen_is_locked())
+    {
+        if (lock_count != 0)
+            mir::fatal_error("Lock count should be zero");
+
+        lock_count++;
+        screen_lock_handle = surface_stack->lock_screen();
+        for (auto const& observer : observers)
+            observer->on_lock();
```

Comment:
This is unsafe if `on_lock` calls either `add_observer` or `remove_observer` (and is UB if other threads call `add_observer` or `remove_observer`, too).

`mir::ObserverMultiplexer<SessionLockObserver>` handles all this.